### ISSUE TITLE
[Bug]: bind_front() and bind_back() fail with gcc 5.5

### ISSUE THREAD

**Issue by pfeatherstone** [davisking/dlib#2869]
### What Operating System(s) are you seeing this problem on?

Linux (x86-64)

### dlib version

master

### Compiler

gcc 5.5

### Expected Behavior

```
#include <cstdio>
#include <type_traits>
#include <dlib/functional.h>

constexpr int test_bind_func(int a, int b)
{
    return a - b;
}

int main()
{
    static_assert(dlib::bind_front(test_bind_func, 50)(3) == 47, "this should be constexpr");
    return dlib::bind_front(test_bind_func, 50)(3);
}
```

### Current Behavior

Does not compiler with gcc 5.5. Fine with gcc 6.1 onward.

### Steps to Reproduce

Run the above.

**Comment by pfeatherstone** [comment]
Potential fix https://github.com/davisking/dlib/pull/2870
Can we setup CI for this? Or have Github Actions deprecated ubuntu 18 with gcc5 ?

**Comment by davisking** [comment]
> Potential fix #2870 Can we setup CI for this? Or have Github Actions deprecated ubuntu 18 with gcc5 ?

Yeah sure, send me a PR that adds something.
